week13/상현/PG_GPS.java

```diff
@@ -0,0 +1,67 @@
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+/**
+ * 실패...
+ * dp + dfs로 시도
+ */
+class Solution {
+    private static List<Integer>[] trees;
+    private static int[][] dp;
+    public int solution(int n, int m, int[][] edge_list, int k, int[] gps_log) {
+        int answer = 0;
+
+        trees = new ArrayList[n + 1];
+        dp = new int[n + 1][k];
+        for (int i = 1; i <= n; i++) {
+            trees[i] = new ArrayList<>();
+            Arrays.fill(dp[i], Integer.MAX_VALUE - k);
+        }
+
+        for (int i = 0; i < m; i++) {
+            trees[edge_list[i][0]].add(edge_list[i][1]);
+            trees[edge_list[i][1]].add(edge_list[i][0]);
+        }
+
+        answer = func(gps_log[0], 0,gps_log, k);
+
+        if (answer == Integer.MAX_VALUE - k) answer = -1;
+        System.out.println(answer);
+        return answer;
+    }
+
+    private static int func(int node, int time, int[] gpsLog, int k) {
+        if (time == k - 1) {
+            if (gpsLog[time] == node) return 0;
+            return dp[node][time];
+        }
+
+        if (dp[node][time] != Integer.MAX_VALUE-k) return dp[node][time];
+
+        int diff = 1;
+        int result = dp[node][time];
+        if (gpsLog[time] == node) diff = 0;
+
+        for (Integer next : trees[node]) {
+            result = Math.min(result, func(next, time + 1, gpsLog, k) + diff);
+        }
+
+        dp[node][time] = result;
+        return dp[node][time];
+    }
```

Comment:
```suggestion
        private static int func(int node, int time, int[] gpsLog, int k) {
        if (time == k - 1) {
            if (gpsLog[time] == node) return 0;
            return dp[node][time] = Integer.MAX_VALUE;
        }

        if (dp[node][time] != -1) return dp[node][time];        

        dp[node][time] = Integer.MAX_VALUE;
        
        int diff = 1;
        if (gpsLog[time] == node) diff = 0;
        
        int result = dp[node][time];

        for (Integer next : trees[node]) {
            if(dp[next][time + 1] == Integer.MAX_VALUE) continue;
            result = Math.min(result, func(next, time + 1, gpsLog, k));
        }

        if(result != Integer.MAX_VALUE) dp[node][time] = result + diff;        
        return dp[node][time];
    }
```
위 함수처럼 변경해준 뒤 위에서 dp배열을 -1로 초기화 해주면 통과됩니다
dp가 이미 계산되었음에도 불구하고 루프가 계속 진행되어 시간초과가 났었네요
살짝만 고쳐서 통과하는 코드라 거의 다 푸시긴 하셨어요!